### ISSUE TITLE
BatchNormalization asserts in evalute mode

### ISSUE THREAD

**Issue by sneiman** [torch/nn#383]
1d BatchNormalization fails with an assert noting the requirement for a 2d tensor - even when not in training mode. The code looks to me like it would work fine - though I have no insight into the nuances of batch norm.

I did submit this before - but it is not showing up as an issue. Sorry for the repeat.

seth

**Comment by sneiman** [comment]
A local hack version demonstrates that the current code will not run with a single input tensor. But I still believe this is needed for using this in real time predictions based on a single input at a time.

**Comment by soumith** [comment]
how about faking a batch size of 1

**Comment by sneiman** [comment]
I am doing that for now – just seems a little funky for real time code in the long run. If it doesn’t make the cut, so be it ;)

This entire message is confidential. If it isn't intended for you, you may not use it – so please throw it away and forget about it.

**Comment by sneiman** [comment]
Hmmm - seems like a chunk of busy work to make this work. The offending bn call is deep in the net, so I have to put a view(1, size) in front of it to solve this problem. However, this doesn’t work during training – which is done with mini–batches of size 16. Seems like I have to build a different net (with the extra view call) and copy parameters from the training version, to make this fly.

Am I missing something that would make this easier? I guess one approach would be to have the net test self.train and modify itself by deleting the appropriate nn.view() call – is this even possible?

S

This entire message is confidential. If it isn't intended for you, you may not use it – so please throw it away and forget about it.

**Comment by albanD** [comment]
@sneiman why only faking the batch size of 1 for this module and not for the entire net? Just add the view(1,size) to the input before forwarding it into the net, am I missing something?

**Comment by sneiman** [comment]
For some nets that might work. But the input dimensionality is important to many layers along the way. Also, any nn.View() layers will effectively reset the dimensionality along the way as well.

This entire message is confidential. If it isn't intended for you, you may not use it – so please throw it away and forget about it.

**Comment by albanD** [comment]
* You can use [this](https://github.com/torch/nn/blob/master/View.lua#L29) function for View to prevent it from reseting the dimensionality.
* If you train your network with a batch size of 16, you use the `forward` method with input with 16 batch size, I don't see why you cannot make it work for input with 1 batch size. Except if you have very different behavior at train and test time for your network.

**Comment by sneiman** [comment]
I struggled with this simple solution for quite some time, as the lua error msgs directly pointed at the 2d v 1d issue. However, the problem was completely different, unrelated to BN(), and now resolved.

This entire message is confidential. If it isn't intended for you, you may not use it – so please throw it away and forget about it.
